presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/TypeMatchingChecker.java

```diff
@@ -0,0 +1,169 @@
+
+import com.facebook.presto.Session;
+import com.facebook.presto.metadata.Metadata;
+import com.facebook.presto.metadata.Signature;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeSignature;
+import com.facebook.presto.sql.parser.SqlParser;
+import com.facebook.presto.sql.planner.SimplePlanVisitor;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.planner.plan.UnionNode;
+import com.facebook.presto.sql.planner.plan.WindowNode;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.QualifiedNameReference;
+import com.google.common.collect.ListMultimap;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.sql.analyzer.ExpressionAnalyzer.getExpressionTypes;
+import static com.facebook.presto.type.UnknownType.UNKNOWN;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Ensures that all the expressions and FunctionCalls matches their output symbols
+ */
+public final class TypeMatchingChecker
+        implements PlanSanityChecker.Checker
+{
+    public TypeMatchingChecker() {}
+
+    @Override
+    public void validate(Session session, Metadata metadata, SqlParser sqlParser, Map<Symbol, Type> types, PlanNode plan)
+    {
+        plan.accept(new Visitor(session, metadata, sqlParser, types), null);
+    }
+
+    private static class Visitor
+            extends SimplePlanVisitor<Void>
+    {
+        private final Session session;
+        private final Metadata metadata;
+        private final SqlParser sqlParser;
+        private final Map<Symbol, Type> types;
+
+        public Visitor(Session session, Metadata metadata, SqlParser sqlParser, Map<Symbol, Type> types)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.sqlParser = requireNonNull(sqlParser, "sqlParser is null");
+            this.types = requireNonNull(types, "types is null");
+        }
+
+        @Override
+        public Void visitAggregation(AggregationNode node, Void context)
+        {
+            visitPlan(node, context);
+
+            AggregationNode.Step step = node.getStep();
+
+            // Do checking only when the step is SINGLE since FINAL and PARTIAL aggregation might have intermediate representation types
+            if (step.equals(AggregationNode.Step.SINGLE)) {
+                checkSignatureTypeMatching(node.getFunctions());
+                checkFunctionCallTypeMatching(node.getAggregations());
+            }
+            else if (step.equals(AggregationNode.Step.FINAL)) {
```

Comment:
@raghavsethi I don't think we could check `node.getFunctions()` for `PARTIAL` since the checking we do here is just trying to match the key of `Map<Symbol, Signature>`(which is returned by `node.getFunctions()`) and the return type of `Signature` while in `PARTIAL` node, the key symbol actually has the type of the intermediate result type of the aggregation function.

E.g. 
```
FINAL:
avg_1(double) --> signature('avg', double as return type)         --> double matches double

PARTIAL:
avg_0(varchar) --> signature('avg', double as return type)        --> varchar doesn't match double
```